xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -228,6 +287,15 @@ void cancelClusterDataWatch(String clusterName, ClusterWatcher watcher) {
     if (watchers.isEmpty()) {
       logger.log(Level.FINE, "Stop watching cluster {0}", clusterName);
       clusterWatchers.remove(clusterName);
+      // Remove the corresponding CDS entry.
+      absentCdsResources.remove(clusterName);
+      clusterNamesToClusterUpdates.remove(clusterName);
```

Comment:
I think this is against the caching behavior in the spec.

>In LDS and CDS, because we need to support legacy management servers that do not honor the resource_names field, we cannot assume that the server will stop sending resources when the client unsubscribes from them.  In this case, the client must use all resources sent by the server, because if it needs to re-subscribe to a resource later, the server will not resend the resource if it has not changed.  In other words, the client must always cache all resources sent by the server, even if the client does not currently have any interest in them.

Reply:
This spec is out-of-date (many parts of the design doc is out dated as changes were made so frequently and the doc was not able to catch up). One major design changed made was:

> The client will silently ignore any supplied resources that were not explicitly requested. When the client sends a new request that changes the set of resources being requested, the server must resend any newly requested resources, even if it previously sent those resources without having been asked for them and the resources have not changed since that time.

(More detail in https://github.com/envoyproxy/envoy/blob/master/api/xds_protocol.rst#how-the-client-specifies-what-resources-to-return)

Therefore, we only need to cache resources that are currently requesting for.

Reply:
Yes, the envoy spec is changed. The question is do we still "support legacy management servers"

>In LDS and CDS, because we need to support legacy management servers ...

Reply:
It seems that "legacy management servers" violate the latest protocol mentioned in envoy's spec. I doubt if the spec for legacy management servers still hold. We need confirmation @markdroth.

Reply:
We do not need to cache any resource that we didn't ask for, because the server is required to resend any previously-sent resource when we send a new request.  This should work fine with legacy management servers (those that don't look at the `resource_names` field for LDS or CDS), because those management servers will always send a new response containing all resources whenever we send a new request.

Reply:
Alright, legacy management servers would always send clients at least the amount of resources that modern management servers do for the same request. This is compatible.

Reply:
@markdroth Then it seems we don't need a cache for CDS at all if we always send a new request whenever we change our subscription, because the management server will then always send a new response. (We may use a cache only for optimization)

Reply:
Cache is needed, for a watcher added subscribing to some resource that is already under watch by some other watcher(s). In this case, no new request is sent, but the newly added watcher gets result from cache.

Reply:
Yes, but that's holding the most current updates for the most current resources subscribed, it's what I meant for optimization. Resources occurred earlier that the most current subscription set don't need be cached anymore.

>Cache is needed, for a watcher added subscribing to some resource that is already under watch by some other watcher(s). In this case, no new request is sent, but the newly added watcher gets result from cache.

Reply:
You're both right: we do not need to cache any resource that the server sent that we did not explicitly request, but we do need to cache any resource that we *did* request, because there is no guarantee that the server will resend it if it has not changed.